docs/3.1/docs/tsa-arch/arch-considerations-upgrading.rst

```diff
@@ -0,0 +1,265 @@
+-------------------
+
+4.1.1 Vetting
+~~~~~~~~~~~~~
+By default, when a Daml package is uploaded, the participant node automatically marks it as vetted and publishes its vetting status on the synchronizer. This allows other participant nodes to determine which workflows the parties on the participant can engage in. A package cannot be used until it is vetted, providing an additional verification step in the deployment process.
+
+4.1.2 Unvetting
+~~~~~~~~~~~~~~~
+Packages can also be unvetted. For example, after uploading and vetting v2, unvetting v1 signals that the participant node can no longer participate in v1 workflows, finalizing the upgrade process. 
+
+Note that all v1 contracts must be fully upgraded before unvetting v1 to avoid potential issues. This requirement extends beyond a zero-downtime upgrade.
+
+4.2 Testing
+-----------
+Once the Daml packages are vetted, it is essential to ensure that the new version remains compatible with the previous versions. This involves two types of testing:
+
+* A type-level compatibility test checks whether the old and new versions of a package with the same name can coexist without breaking. The easiest way to test this is by uploading both old and new versions to a fresh participant node as part of CI. To do this, access the DAR files used in production. Ideally, these should be stored in a dedicated artifact repository, but given their small sizes (typically under 1 MB), they may also be checked into source control.
+
+* A workflow-level compatibility test verifies that core business processes (workflows) continue to function correctly after an upgrade. At a minimum, it is recommended to include one integration test. A basic integration test should follow these steps:
+
+  1. Start the application with v2 software, but upload only the v1 DAR file to test backward compatibility.
+  2. Initialize the application and start one instance of every core workflow.
+  3. Upload the v2 DAR.
+  4. Update the configuration to instruct the backends to start using the v2 DAR.
+  5. Verify that the workflows remain in the correct state and can continue without issues. 
+
+For more complex upgrades, additional tests may be needed. 
+
+4.3 Rollback
+------------
+
+4.3.1 Rollback by Unvetting
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+To roll back upgrades that do not modify the types of existing templates and choices, unvet the v2 DAR package.
+
+4.3.2 Rollback by Roll-Forward
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For upgrades that add new fields to existing templates, rollback becomes more complex. In such cases, the rollback must be performed in a “roll-forward” fashion by publishing a new upgrade. This is necessary because if at least one contract has been created using the new fields, those contracts cannot be read with the previous version of the Daml code. Simply unvetting v2 is not an option unless it is acceptable for contracts using the new fields to no longer be referenced. Instead, follow these steps:
```

Comment:
```suggestion
Rollback is more complex for upgrades that add new fields to existing templates. In such cases, the rollback must be performed as a “roll-forward” by publishing a new upgrade. This is necessary because if at least one contract has been created using the new fields, those contracts cannot be read with the previous version of the Daml code. Unvetting v2 is not an option unless it is acceptable for contracts using the new fields to no longer be referenced. Instead, follow these steps:
```